provider/oci/environ.go

```diff
@@ -0,0 +1,167 @@
+// Copyright 2018 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package oci
+
+import (
+	"sync"
+
+	"github.com/juju/errors"
+	"github.com/juju/utils/clock"
+	"github.com/juju/version"
+
+	"github.com/juju/juju/constraints"
+	"github.com/juju/juju/environs"
+	"github.com/juju/juju/environs/config"
+	"github.com/juju/juju/environs/instances"
+	"github.com/juju/juju/instance"
+	"github.com/juju/juju/provider/common"
+	providerCommon "github.com/juju/juju/provider/oci/common"
+	"github.com/juju/juju/storage"
+
+	ociCore "github.com/oracle/oci-go-sdk/core"
+)
+
+type Environ struct {
+	cli       providerCommon.ApiClient
+	p         *EnvironProvider
+	clock     clock.Clock
+	ecfgMutex sync.Mutex
+	ecfgObj   *environConfig
+	namespace instance.Namespace
+
+	vcn     ociCore.Vcn
+	seclist ociCore.SecurityList
+	// subnets contains one subnet for each availability domain
+	// these will get created once the environment is spun up, and
+	// will never change.
+	subnets map[string][]ociCore.Subnet
+}
+
+var _ common.ZonedEnviron = (*Environ)(nil)
+var _ storage.ProviderRegistry = (*Environ)(nil)
+var _ environs.Environ = (*Environ)(nil)
+var _ environs.Firewaller = (*Environ)(nil)
+var _ environs.Networking = (*Environ)(nil)
+var _ environs.NetworkingEnviron = (*Environ)(nil)
+
+func (e *Environ) ecfg() *environConfig {
+	e.ecfgMutex.Lock()
+	defer e.ecfgMutex.Unlock()
+	return e.ecfgObj
```

Comment:
Sigh,
digging into why this has to exist, and why we might want to update the config late is all down to a single key that we inject ("agent-version") which isn't even relevant to the Environ itself (unless it is setting something like user-agent headers in requests.)